### DISCUSSION TITLE
Best effort mapping from GIDs to Unicode codepoints

### DISCUSSION THREAD

**Discussion by ceztko** [harfbuzz/harfbuzz#3267] [Q&A] [answered]
Hello, I have some already shaped text that comes in form of raw list of GIDs mapping to a OpenType font, without semantical Unicode representation. I would like to implement a best effort functionality to map these GIDs back to Unicode codepoints, well aware that this is not doable in general, because of some languages using ligatures, contextual glyph substitution, Right To Left text conventions or a combinations of these. I know that the map from Unicode code point to GID should be present in the `cmap` table of the OpenType font (I will just ignore any glyph substitution rule present in the `gsub` table, if present). The question is: can HarfBuzz aid me in any way to implement this functionality or I should just read the `cmap` table in the font directly? I know that HarbBuzz has [`hb_face_collect_unicodes` ](https://harfbuzz.github.io/harfbuzz-hb-face.html#hb-face-collect-unicodes) but that seems only to collect the codes in a set, and have no notion of the original glyph indices. I'm very new to HarfBuzz, I will certainly use it to do some text shaping later but for now that's the task I have to do.

**Comment by khaledhosny** [comment, marked as answer]
If you want to avoid parsing `cmap` table yourself, you can use [`hb_face_collect_unicodes()`](https://harfbuzz.github.io/harfbuzz-hb-face.html#hb-face-collect-unicodes) to collect all Unicode code points supported by the font, then [`hb_font_get_nominal_glyphs()`](https://harfbuzz.github.io/harfbuzz-hb-font.html#hb-font-get-nominal-glyphs) to map them to glyph IDs and you can then build a reverse map yourself.

However, even if you limit yourself to reversing `cmap` mapping, there is still the possibility of multiple code points mapping to the same glyph, and Unicode variation selectors where two code points map to one glyph.

**Reply by khaledhosny** [reply]
Some applications, e.g. PDF readers, will try to use glyph names instead, following [Adobe Glyph List Specification](https://github.com/adobe-type-tools/agl-specification) but this only works when fonts follow that naming scheme (which is a common convention, but not mandatory).

**Reply by ceztko** [reply]
Thank you a lot! Yes, I was also aware of the other pitfalls you mentioned. This is supposed to be a pure best-effort functionality, and the user will be informed if the mapping fails but the end result will at least pass some other formal validations.

**Reply by behdad** [reply]
Another trick PDF readers do, which you can also do using HarfBuzz, is to get glyph name (`hb_font_get_glyph_name`), trim anything after a `.`, if it starts with `uni`, parse the rest as a Unicode value hex number.
